src/overmap.cpp

```diff
@@ -4397,11 +4483,81 @@ void overmap::open()
         for (int i = -1; i <= 1; i += 2) {
             pointers.push_back(overmap_buffer.get_existing(loc.x+i, loc.y));
         }
+
         // pointers looks like (north, south, west, east)
-        generate(pointers[0], pointers[3], pointers[1], pointers[2]);
+        generate_outer( pointers[0], pointers[3], pointers[1], pointers[2] );
     }
 }
 
+void overmap::generate_outer( const overmap* north, const overmap* east, const overmap* south, const overmap* west )
+{
+    // This string is here because it is long and we don't want indents eating up precious space
+    static const std::string menu_s = _(
+"Couldn't generate overmap with current settings.\n"
+"The following specials could not be placed:\n%s");
+    current_validity = overmap_valid::invalid;
+    do {
+        int try_num = std::max( 1, get_option<int>( "OVERMAP_GENERATION_TRIES" ) );
+        for( int i = 0; i < try_num && current_validity != overmap_valid::valid; i++ ) {
+            clear();
+            generate( north, east, south, west );
+        }
+
+        // Note: this should not be checked in a loop - only totally valid maps should exit the loop early
+        if( current_validity >= allow_generation ) {
+            // Good enough
+            break;
+        }
+
+        const auto &opt = get_option<std::string>( "ALLOW_INVALID_OVERMAPS" );
+        if( g != nullptr && ( opt == "ask_invalid" || opt == "ask_unlimited" ) ) {
+            uimenu askmenu;
```

Comment:
That's an UI thing. I'd suggest moving it out of `overmap` (e.g. somewhere into the `game_menus` namespace).